### ISSUE TITLE
[RFC] [4.0] Standard for naming of classes?

### ISSUE THREAD

**Issue by ciar4n** [joomla/joomla-cms#15279]
We could do with a standard for naming of class attributes. Maybe one already exists? 

Understandably naming is currently quite random..

https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/edit_assignment.php#L62
https://github.com/joomla/joomla-cms/blob/staging/modules/mod_languages/tmpl/default.php#L19 https://github.com/joomla/joomla-cms/blob/staging/modules/mod_syndicate/tmpl/default.php#L13 https://github.com/joomla/joomla-cms/blob/staging/modules/mod_tags_popular/tmpl/default.php#L14

Some mainstream standards exist.. most notable is probably BEM... https://en.bem.info/

**Comment by brianteeman** [comment]
I was at a presentation at a php conference on this exact topic last weekend

You can see the slidedeck here https://drive.google.com/file/d/0B7FGMFd-aDPWQlpMaDdna2ExbWM/view

I hope the video will be online in a few days

**Comment by ciar4n** [comment]
There is likely to be some principles of PHP naming that could be used here. I guess HTML classes would have less variables which would allow us more specific naming.

BEM (Block, Element, Modifier) is effective. However I am not sure how we feel about using double underscores and hyphens.

The BEM principle..

```
/* Block component */
.btn {}

/* Element that depends upon the block */ 
.btn__login {}

/* Modifier that changes the style of the block */
.btn_orange {} 
.btn_big {}
```

One of its benefits is it allows everything to have a unique class which makes CSS specificity very flat and low (less nesting). You can also tell how a class can be reused without conflict and its function by the class name (Eg. btn--orange - a modifier only used within any btn).

The flip side is some think it makes for ugly looking code. I'm not suggesting we go to the extreme of BEM but maybe we can borrow from it.

How we name the 'blocks' is something to consider separately. For the highest level blocks we could use the Joomla architecture...

`ExtensionType-Extension-View` (Eg. `com-content-archive`).

I am not sure how sustainable this is throughout all the markup, anyone with a broader knowledge of Joomla architecture could probably answer this. My knowledge here has proven to be limited at best (",)

**Comment by yvesh** [comment]
@ciar4n hmph, not a sure if i am a fan of double hyphens or underscores.. Is it really making things easier? 

Yes, you can easier see the relations and dependencies, but i think most know from Bootstrap that you have a .btn as base and a .btn-primary or an alert and alert-info etc.

**Comment by ciar4n** [comment]
> @ciar4n hmph, not a sure if i am a fan of double hyphens or underscores

I'm not a huge fan either. It just avoids over nesting in CSS/SCSS. So instead of ..
```
.block-name {
  .element-name {
    ...
  }
}
```

You use the following which makes for better CSS specificity (will never conflict)..

```
.block-name__element-name {
  ...
}
```

It is not pretty and I am in no way suggesting we use it. Just thought I would mention it as it has its merits and is considered a gold standard by many (https://github.com/airbnb/css).

> Maybe as surround class? (Strapping)

Strapping I think is fine and should be expected. We just need some sort of naming convention for the elements that has an application for every possible case.

**Comment by laoneo** [comment]
I started to use BEM in my extensions and can say that it makes the markup consistent. The good thing is that it gives you a clear simple guidance. Perhaps somebody should come up with a small pr on a back end view to illustrate the usage of it in Joomla.

**Comment by asika32764** [comment]
I'm using Airbnb style and they recommends to use BEM and OOCSS. I'm more love OOCSS.

https://github.com/airbnb/css

The BS4 seems follows this style and OOCSS and always keep nested structure less than 3 levels.

**Comment by dgrammatiko** [comment]
So how BEM will work with Bootstrap? Are you going to rewrite BS4 to BEM?

**Comment by laoneo** [comment]
One does not exclude another, if we use the BS4 classes then we have to mix them. Or do I miss here something?

**Comment by ciar4n** [comment]
> One does not exclude another, if we use the BS4 classes then we have to mix them. 

I agree. I think we should have both. Bootstrap classes are generally utility classes, or in BEM terms 'modifiers'. Bootstrap does not include unique identifiers for each element which I believe we need. BEM is a naming convention that would allow us to do this. 

Currently we tend to use Bootstrap classes as the identifiers, which I believe is incorrect and also couples Joomla even tighter with Bootstrap. Say for example you wish to style a module login button you currently would have something like..

```
.mod-login .btn.btn-primary {
   ...
}
```

The above is over complex, tightly coupled to Bootstrap, nesting, and has little specificity (presumes there is only one `.btn.btn-primary`

Better to have something like..

```
.mod-login__btn--login {
   ...
}
```

**Comment by ciar4n** [comment]
As I see it here are some achievable goals we can expect from a naming convention...

- A strict naming convention for HTML classes adoptable throughout Joomla. 

- All elements are uniquely identifiable and can be selected within CSS. This excludes framework classes (Bootstrap). This does not mean that every element needs to have class. In most cases elements can be selected by the class of a parent element.

- The root element (containing div) of a view can be easily identified within the DOM.

- The location of a view within the Joomla architecture can be easily identified by the root element of a views class name (`ExtensionType-Extension-View` (Eg. `com-content-archive`).

**Comment by ghost** [comment]
closed as having Pull Request #20147
